### PR TITLE
[SuperEditor] Convert task to paragraph on BACKSPACE (Resolves #1283)

### LINKED ISSUE #1283: [SuperEditor] In the latest main branch, deleting a Task node causes an exception
In the latest main branch, deleting a Task node causes an exception, and then the entire editor becomes unresponsive.

SuperEditor: main

flutter doctor:
[✓] Flutter (Channel master, 3.13.0-14.0.pre.5, on macOS 13.4.1 22F770820d darwin-arm64, locale zh-Hans-CN)
[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2022.3)
[✓] IntelliJ IDEA Ultimate Edition (version 2022.3.3)
[✓] VS Code (version 1.81.0)

 Video:

https://github.com/superlistapp/super_editor/assets/37604230/c5d3717c-1585-4ed9-9253-d6067f448445

Error:
```
======== Exception caught by services library ======================================================
The following _Exception was thrown during method call TextInputClient.updateEditingStateWithDeltas:
Exception: Could not handle EditorRequest. DocumentEditor doesn't have a handler that recognizes the request: Instance of 'DeleteUpstreamAtBeginningOfNodeRequest'

When the exception was thrown, this was the stack: 
#0      Editor._findCommandForRequest (package:super_editor/src/core/editor.dart:189:5)
#1      Editor.execute (package:super_editor/src/core/editor.dart:147:23)
#2      TextDeltasDocumentEditor.delete (package:super_editor/src/default_editor/document_ime/document_delta_editing.dart:370:14)
#3      TextDeltasDocumentEditor._applyDeletion (package:super_editor/src/default_editor/document_ime/document_delta_editing.dart:219:5)
#4      TextDeltasDocumentEditor.applyDeltas (package:super_editor/src/default_editor/document_ime/document_delta_editing.dart:83:9)
#5      DocumentImeInputClient.updateEditingValueWithDeltas (package:super_editor/src/default_editor/document_ime/document_ime_communication.dart:196:30)
#6      DeltaTextInputClientDecorator.updateEditingValueWithDeltas (package:super_editor/src/default_editor/document_ime/ime_decoration.dart:129:14)
#7      TextInput._handleTextInputInvocation (package:flutter/src/services/text_input.dart:1864:63)
#8      TextInput._loudlyHandleTextInputInvocation (package:flutter/src/services/text_input.dart:1753:20)
#9      MethodChannel._handleAsMethodCall (package:flutter/src/services/platform_channel.dart:552:55)
#10     MethodChannel.setMethodCallHandler.<anonymous closure> (package:flutter/src/services/platform_channel.dart:545:34)
#11     _DefaultBinaryMessenger.setMessageHandler.<anonymous closure> (package:flutter/src/services/binding.dart:567:35)
#12     _invoke2 (dart:ui/hooks.dart:202:13)
#13     _ChannelCallbackRecord.invoke (dart:ui/channel_buffers.dart:45:5)
#14     _Channel.push (dart:ui/channel_buffers.dart:135:31)
#15     ChannelBuffers.push (dart:ui/channel_buffers.dart:343:17)
#16     PlatformDispatcher._dispatchPlatformMessage (dart:ui/platform_dispatcher.dart:745:22)
#17     _dispatchPlatformMessage (dart:ui/hooks.dart:114:31)
call: MethodCall(TextInputClient.updateEditingStateWithDeltas, [2, {deltas: [{selectionBase: 1, oldText: . Cmmatic control over the user's selection and styles, create a DocumentComposer., selectionAffinity: TextAffinity.downstream, deltaEnd: -1, deltaText: , deltaStart: -1, composingExtent: -1, selectionIsDirectional: false, selectionExtent: 2, composingBase: -1}, {selectionBase: 1, oldText: . Cmmatic control over the user's selection and styles, create a DocumentComposer., selectionAffinity: TextAffinity.downstream, deltaEnd: 2, deltaText: , deltaStart: 1, composingExtent: -1, selectionIsDirectional: false, selectionExtent: 1, composingBase: -1}]}])
====================================================================================================
```

### PR DESCRIPTION
**Pull request by angelosilvestre**
[SuperEditor] Convert task to paragraph on BACKSPACE. Resolves #1283.

Pressing backspace at the beginning of a task causes the editor to crash with the exception:

```console
Exception: Could not handle EditorRequest. DocumentEditor doesn't have a handler that recognizes the request: Instance of 'DeleteUpstreamAtBeginningOfNodeRequest'
```

The cause is that we don't have a handler registered to handle the `DeleteUpstreamAtBeginningOfNodeRequest` with a `TaskNode`.

This PR introduces a `ConvertTaskToParagraphCommand` to convert a task to a paragraph when pressing backspace at the beginning of the task.

### CODE REVIEW

**Comment by matthew-carroll** [comment]
@angelosilvestre is this ready for review?

**Comment by angelosilvestre** [comment]
@matthew-carroll Yeah. Forgot to request a review

File: super_editor/lib/src/default_editor/default_document_editor.dart [hunk 1]
```diff
@@ -147,6 +150,12 @@ final defaultRequestHandlers = List.unmodifiable(<EditRequestHandler>[
           newNodeId: request.newNodeId,
         )
       : null,
+  (request) => request is ConvertTaskToParagraphRequest //
```
Comment (matthew-carroll) [inline super_editor/lib/src/default_editor/default_document_editor.dart:153]: Wouldn't it make more sense to put this up near `ConvertParagraphToTaskRequest` and/or move all the task requests near each other?

File: super_editor/lib/src/default_editor/tasks.dart [hunk 2]
```diff
@@ -399,6 +438,47 @@ class ConvertParagraphToTaskCommand implements EditCommand {
   }
 }
 
+class ConvertTaskToParagraphRequest implements EditRequest {
+  ConvertTaskToParagraphRequest({
```
Comment (matthew-carroll) [inline super_editor/lib/src/default_editor/tasks.dart:442]: Make constructor a `const`, add a blank line after the constructor, and generate the hashcode and equals methods.

File: super_editor/lib/src/default_editor/tasks.dart [hunk 3]
```diff
@@ -306,6 +306,45 @@ ExecutionInstruction enterToInsertNewTask({
...
+    return ExecutionInstruction.continueExecution;
+  }
+
+  if ((editContext.composer.selection!.extent.nodePosition as TextPosition).offset > 0) {
+    // The selection isn't at the beginning.
+    return ExecutionInstruction.continueExecution;
+  }
+
+  editContext.editor.execute([
+    ConvertTaskToParagraphRequest(
```
Comment (matthew-carroll) [inline super_editor/lib/src/default_editor/tasks.dart:339]: Shouldn't this send a request to delete at the beginning of a node?

File: super_editor/lib/src/default_editor/default_document_editor.dart (hunk 1 above)
Comment (angelosilvestre) [reply super_editor/lib/src/default_editor/default_document_editor.dart:153]: Moved all requests close to each other.

File: super_editor/lib/src/default_editor/tasks.dart (hunk 3 above)
Comment (matthew-carroll) [reply super_editor/lib/src/default_editor/tasks.dart:340]: The concern that I have is that what if an app wants to handle that beginning deletion differently? We've introduced the concept of a deletion at the beginning, right? And I assume that's what would be triggered by an equivalent IME action, right? But this key handler does something different than IME, and also takes away the opportunity for apps to override what happens when deletion happens at the beginning of a task node.

Am I missing anything?

File: super_editor/lib/src/default_editor/tasks.dart [hunk 4]
```diff
@@ -399,6 +435,38 @@ class ConvertParagraphToTaskCommand implements EditCommand {
   }
 }
 
+class ConvertTaskToParagraphCommand implements EditCommand {
+  ConvertTaskToParagraphCommand({
```
Comment (matthew-carroll) [inline super_editor/lib/src/default_editor/tasks.dart:439]: Can this be `const`?

File: super_editor/lib/src/default_editor/tasks.dart [hunk 5]
```diff
@@ -399,6 +435,38 @@ class ConvertParagraphToTaskCommand implements EditCommand {
...
+    final newParagraphNode = ParagraphNode(
+      id: taskNode.id,
+      text: taskNode.text,
+      metadata: newMetadata,
+    );
+    document.replaceNode(oldNode: taskNode, newNode: newParagraphNode);
+
+    executor.logChanges([
+      DocumentEdit(
+        NodeChangeEvent(taskNode.id),
```
Comment (matthew-carroll) [inline super_editor/lib/src/default_editor/tasks.dart:464]: Is this the same event that we use in other similar situations?

Comment (angelosilvestre) [reply super_editor/lib/src/default_editor/tasks.dart:464]: Yeah, this is the same event logged by `ConvertListItemToParagraphCommand`.

**Review by matthew-carroll — APPROVED** [review]
LGTM

**Pull request merged**